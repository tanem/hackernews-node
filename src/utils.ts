import jwt from 'jsonwebtoken'
import { Context } from './context'

export const APP_SECRET = 'GraphQL-is-aw3some'

export const getUserId = (context: Context) => {
  const Authorization = context.request.get('Authorization')
  if (Authorization) {
    const token = Authorization.replace('Bearer ', '')
    const { userId } = jwt.verify(token, APP_SECRET) as { userId: number }
    return userId
  }

  throw new Error('Not authenticated')
}
