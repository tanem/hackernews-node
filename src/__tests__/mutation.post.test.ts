import bcrypt from 'bcryptjs'
import Chance from 'chance'
import jwt from 'jsonwebtoken'
import path from 'path'
import request from 'supertest'
import { APP_SECRET } from '../utils'

const chance = Chance(path.basename(__filename))

let httpServerUrl: typeof global.__HTTP_SERVER_URL__
let prisma: typeof global.__PRISMA__
let token: string

beforeAll(async () => {
  httpServerUrl = global.__HTTP_SERVER_URL__
  prisma = global.__PRISMA__
})

const name = chance.first()
const email = chance.email()
const password = chance.word()

beforeEach(async () => {
  const { id: userId } = await prisma.user.create({
    data: {
      name,
      email,
      password: await bcrypt.hash(password, 10),
    },
  })
  token = jwt.sign({ userId }, APP_SECRET)
})

test('successful', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .set('Authorization', `Bearer ${token}`)
    .send({
      query: `
      mutation {
        post(
          url: "${chance.url()}"
          description: "${chance.sentence({ words: 5 })}"
        ) {
          id
          description
          url
          postedBy {
            id
            name
          }
          votes {
            id
            user {
              id
              name
            }
          }
        }
      }
    `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})

test('returns error if not authenticated', async () => {
  const { body, status } = await request(httpServerUrl)
    .post('/')
    .send({
      query: `
      mutation {
        post(
          url: "${chance.url()}"
          description: "${chance.sentence({ words: 5 })}"
        ) {
          id
          description
          url
          postedBy {
            id
            name
          }
          votes {
            id
            user {
              id
              name
            }
          }
        }
      }
    `,
    })

  expect(status).toBe(200)
  expect(body).toMatchSnapshot()
})
