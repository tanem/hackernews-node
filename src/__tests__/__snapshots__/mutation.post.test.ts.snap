// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns error if not authenticated 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "locations": Array [
        Object {
          "column": 9,
          "line": 3,
        },
      ],
      "message": "Not authenticated",
      "path": Array [
        "post",
      ],
    },
  ],
}
`;

exports[`successful 1`] = `
Object {
  "data": Object {
    "post": Object {
      "description": "Cermamu tidrucnil nulus javwojzov dukkahkor.",
      "id": 1,
      "postedBy": Object {
        "id": 1,
        "name": "Josephine",
      },
      "url": "http://iloemfa.de/magila",
      "votes": Array [],
    },
  },
}
`;
